File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/interactors/DecksInteractor.java
package com.dbottillo.mtgsearchfree.interactors;


import android.net.Uri;

import com.dbottillo.mtgsearchfree.model.Deck;
import com.dbottillo.mtgsearchfree.model.MTGCard;

import java.util.List;

import rx.Observable;

public interface DecksInteractor {

    Observable<List<Deck>> load();

    Observable<List<MTGCard>> loadDeck(Deck deck);

    Observable<List<Deck>> addDeck(String name);

    Observable<List<Deck>> deleteDeck(Deck deck);

    Observable<List<MTGCard>> editDeck(Deck deck, String name);

    Observable<List<MTGCard>> addCard(String name, MTGCard card, int quantity);

    Observable<List<MTGCard>> addCard(Deck deck, MTGCard card, int quantity);

    Observable<List<MTGCard>> removeCard(Deck deck, MTGCard card);

    Observable<List<MTGCard>> removeAllCard(Deck deck, MTGCard card);

    Observable<List<Deck>> importDeck(Uri uri);
}

File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/interactors/DecksInteractor.kt
package com.dbottillo.mtgsearchfree.interactors

import android.net.Uri
import com.dbottillo.mtgsearchfree.model.CardsCollection
import com.dbottillo.mtgsearchfree.model.Deck
import com.dbottillo.mtgsearchfree.model.DeckCollection
import com.dbottillo.mtgsearchfree.model.MTGCard
import io.reactivex.Observable
import io.reactivex.Single

interface DecksInteractor {
    fun load(): Single<List<Deck>>
    fun loadDeck(deck: Deck): Observable<DeckCollection>
    fun addDeck(name: String): Observable<List<Deck>>
    fun deleteDeck(deck: Deck): Observable<List<Deck>>
    fun editDeck(deck: Deck, name: String): Observable<DeckCollection>
    fun addCard(name: String, card: MTGCard, quantity: Int): Observable<DeckCollection>
    fun addCard(deck: Deck, card: MTGCard, quantity: Int): Observable<DeckCollection>
    fun removeCard(deck: Deck, card: MTGCard): Observable<DeckCollection>
    fun removeAllCard(deck: Deck, card: MTGCard): Observable<DeckCollection>
    fun importDeck(uri: Uri): Observable<List<Deck>>
    fun exportDeck(deck: Deck, cards: CardsCollection): Observable<Boolean>
    fun moveCardToSideboard(deck: Deck, card: MTGCard, quantity: Int): Observable<DeckCollection>
    fun moveCardFromSideboard(deck: Deck, card: MTGCard, quantity: Int): Observable<DeckCollection>
}